### DISCUSSION TITLE
View collection permissions

### DISCUSSION THREAD

**Discussion by LickABrick** [pocketbase/pocketbase#4041] [Q&A] [answered]
Hi, I am a little confused on how View collection permissions work. Do view collection permissions look at permissions set on normal collections too?

In my app a users can be part of multiple (or a single) tenant. The tenants a user is part of is stored in the users collection (as a relation). How can I show all users.name and tenant.name from the tenant(s) the user is linked to without showing any other users?

**Comment by ganigeorgiev** [comment]
Without details about your collections structure it is hard to understand what you are asking and what actually is the issue.

In general, View collection's List and View API rules are independent from the API rules of the collection(s) that the view query is derived from (_note that it may not be even from a collection_). Or in other words, they should behave the same as any other collection type.

**Reply by LickABrick** [reply]
Here is the structure of the relevant collections:

Users:
![image](https://github.com/pocketbase/pocketbase/assets/17205583/1a80b00d-322f-40ca-8c65-1f1cb576a60a)

![image](https://github.com/pocketbase/pocketbase/assets/17205583/22566d30-7204-414c-b71f-93557f464c81)

Tenants:
![image](https://github.com/pocketbase/pocketbase/assets/17205583/608ff409-faa2-4700-9fee-4bc9a82de581)

![image](https://github.com/pocketbase/pocketbase/assets/17205583/922538dc-2873-4726-8f5a-3c3c1d1e0be6)

TenantUsers:
![image](https://github.com/pocketbase/pocketbase/assets/17205583/f6705900-5fa8-4078-b4f3-91b20586161a)
(no permissions set yet)

When a user lists the TenantUsers view collection I only want them to be able to see users who are part of a Tenant they are also a part of.

Sorry if this sounds confusing. If you need any more information let me know

**Reply by ganigeorgiev** [reply]
> When a user lists the TenantUsers view collection I only want them to be able to see users who are part of a Tenant they are also a part of.

If you add the `users.tenants` fields in the view query, you should be able to use an API rule like:

```js
// is authenticated
@request.auth.id != "" &&
// has at least one existing tenant
tenants:length > 0 &&
// has at least one tenant matching with the authenticated user tenants
tenants.id ?= @request.auth.tenants.id
```

But why do you need a view for this? Unless you want to restrict the access of "users" fields, you should be able to apply the above rule directly in the "users" collections without the view collection.

**Reply by LickABrick** [reply]
I do want to restrict access to the users collection as this contains private information about the user.

I have thought about using a separate usersPrivate collection with the private information but thought using a view collection was a better solution.

**Reply by ganigeorgiev** [reply]
In this case I guess the view collection could be used but you'll have to add the `tenants` field in the query so that you can target it in the API rules.

**Reply by ganigeorgiev** [reply, marked as answer]
Another alternative if you don't want the tenants list to be visible could be to have a view query like:

```sql
SELECT
    users.id,
    users.name,
    users.avatar,
    users.id as user -- add the id again so that it can be used as a relation field
FROM users
```
And then the view API rules could look like:

```js
// is authenticated
@request.auth.id != "" &&
// has at least one existing tenant
user.tenants:length > 0 &&
// has at least one tenant matching with the authenticated user tenants
user.tenants.id ?= @request.auth.tenants.id
```
(_alternative of the extra "user" relation field could be to manually join based on the view id using `@collection.users.id`_)

**Reply by LickABrick** [reply]
Thanks, that works!
